### PR TITLE
CHAT-2289 Add Long Polling HTTP socket in WebRTC lib

### PR DESCRIPTION
**Pull request by DinhNguyenCat**
/cc: @dscherba 

Hi David,
This PR contains the following information:
- Merge the code from PR https://github.com/PersonifyInc/webrtc/pull/3 to new branch upstream-speex.
- Update based on your comment in that PR.
- Update code to support connect with tls in case of XMPP and HTTPS.
- Major update: Introduce the backbone of the HTTP Long Polling socket (XEP124) written in the new class BoshSocket.This class can support main functionality of long polling socket for HTTP and HTTPS.

The code is test passed not under proxy server. Not yet tested under proxy server.
Could you please help to review this PR and provide me your feedback about the idea so that I can update my code.

Things to do next:
- Implement the proxy detection in webrtc, I've seen it in webrtc code but still don't know how to apply it into our environment.
- The BoshSocket contains some code that execute parallel but the code still didn't support thread-safe.
- Other functionality of BoshSocket: inactivity, overactivity handler... is still not supported yet.

Thanks,
Dinh Nguyen

### CODE REVIEW

**Comment by dscherba** [comment]
@DinhNguyenCat : thanks this update. Will comment on the implementation/design shortly.

One behavioral note: this code does support having (and using) two sockets/HTTP connections [good], but it's still not keeping a channel open for server-initiated signaling.

Take a look at XEP-0124's sequence diagrams, especially the part about empty <body> requests. The idea is that the client can always keep a request pending on the server so the server can reply (e.g. if an invite comes in). This implementation will still wait around until the next client event (e.g. an XMPP ping) to get that server response and this wait degrades the user experience.

**Comment by dscherba** [comment]
@DinhNguyenCat : added some comments. Also see https://github.com/PersonifyInc/Network/blob/master/src/Connection.cpp for an example of how to do proxy stuff in a way that works (auto-detection, connection hoops that you'll need to jump through, some HTTP tweaks)

**Comment by DinhNguyenCat** [comment]
@dscherba 

I've just updated the webrtc code with the following updates:
- Update code based on your last comment.
  About your comment about using other flags in stead of port number, currently I still didn't figure out the suitable flag for this, so temporary I set a const port value to avoid hard code in those checks.
- Using proxy detection implemented in webrtc to do the proxy detect functionality for connecting under proxy server.

This is tested passed by me under proxy server in viet nam side (both HTTP and HTTPS).
Could you please help taking a look at this PR and provide me your feedback.

P/s: There is still an error when start call, relating to the RTMP connection. I've sent a email about this problem. But generally, the XMPP connection is ok now and we can start chatting successfully.

Thanks,
Dinh Nguyen

**Comment by DinhNguyenCat** [comment]
/cc: @dscherba @willkelleher 

Please help to review this PR. The update contains:
- Update code based on the last comments.
- Adding code to handle proxy detection and connect through proxy.

P/S: As discussion, about the optimization in BoshSocket (empty Body etc..). I will address it in another task and update it after the release for Lenovo finished.

Thanks,
Dinh Nguyen

**Comment by DinhNguyenCat** [comment]
Ping @dscherba @willkelleher 

@dscherba : I've just squash the 2nd commit into 1st. Thank you for your advice.

File: webrtc/libjingle/xmpp/xmppclient.cc [hunk 1]
```diff
@@ -136,17 +137,26 @@ XmppReturnStatus XmppClient::Connect(
...
+  // Set connection socket server
+  d_->engine_->SetSocketServer(settings.server().hostname(), (int)settings.server().port());
+
   d_->pass_ = settings.pass();
   d_->auth_mechanism_ = settings.auth_mechanism();
   d_->auth_token_ = settings.auth_token();
   d_->server_ = settings.server();
-  d_->proxy_host_ = settings.proxy_host();
-  d_->proxy_port_ = settings.proxy_port();
+  d_->proxy_host_ = (!proxyInfo.address.hostname().empty()) ? proxyInfo.address.hostname(): settings.proxy_host();
```
Comment (dscherba) [inline webrtc/libjingle/xmpp/xmppclient.cc]: trinary operator spacing: (cond) ? if-true : if-false;

File: webrtc/libjingle/xmpp/xmppengineimpl.h [hunk 2]
```diff
@@ -247,6 +273,8 @@ class XmppEngineImpl : public XmppEngine {
   XmppOutputHandler* output_handler_;
   XmppSessionHandler* session_handler_;
 
+  XmppStanzaGenerator* content_wrapper_;
```
Comment (dscherba) [inline webrtc/libjingle/xmpp/xmppengineimpl.h]: I think we should change content_wrapper_ to stanza_generator_ or something (after we changed the class name, this older name doesn't make much sense)

File: webrtc/libjingle/xmpp/boshsocket.cc [hunk 3]
```diff
@@ -0,0 +1,444 @@
...
+    delete secondary_socket_;
+  }
+
+#ifndef USE_SSLSTREAM
+  void BoshSocket::OnReadEvent(rtc::AsyncSocket *socket) { 
+    buzz::InputMsg input_msg;
+    char bytes[4096];
+    input_msg.len_ = socket->Recv(bytes, sizeof(bytes));
+    if (input_msg.len_ > 0) {
+      input_msg.data_ = new char[input_msg.len_];
```
Comment (dscherba) [inline webrtc/libjingle/xmpp/boshsocket.cc:102]: is this deleted anywhere?

File: webrtc/libjingle/xmpp/boshsocket.cc [hunk 4]
```diff
@@ -0,0 +1,444 @@
...
+        }
+      }
+    }
+    if ((events & rtc::SE_READ)) {
+      buzz::InputMsg input_msg;
+      char bytes[4096];
+      size_t read;
+      rtc::StreamResult result = stream->Read(bytes, sizeof(bytes), &read, NULL);
+      if (result == rtc::SR_SUCCESS) {
+        input_msg.data_ = new char[read];
```
Comment (dscherba) [inline webrtc/libjingle/xmpp/boshsocket.cc:229]: deleted anywhere?

File: webrtc/libjingle/xmpp/boshsocket.h [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+      connect_sent_ = false;
+      state_ = buzz::AsyncSocket::State::STATE_CLOSED;
+    }
+  } ManagedSocket;
+
+  typedef struct OutputMsg{
+    char *data_;
+    size_t len_;
+    OutputMsg(const char *str, size_t len) {
+      data_ = new char[len];
```
Comment (dscherba) [inline webrtc/libjingle/xmpp/boshsocket.h:63]: deleted anywhere?

File: webrtc/libjingle/xmpp/boshsocket.h [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+      strcpy(data_, str);
```
Comment (dscherba) [inline webrtc/libjingle/xmpp/boshsocket.h]: if the len is accurate, memcpy might be better -- strcpy() has some security implications. Also should make sure we have guards around so errant signaling can't create buffer issues.

File: webrtc/libjingle/xmpp/xmppclient.cc (hunk 1 above)
Comment (DinhNguyenCat) [inline webrtc/libjingle/xmpp/xmppclient.cc]: Noted, will update

File: webrtc/libjingle/xmpp/xmppengineimpl.h (hunk 2 above)
Comment (DinhNguyenCat) [inline webrtc/libjingle/xmpp/xmppengineimpl.h]: Noted, will update

File: webrtc/libjingle/xmpp/boshsocket.cc (hunk 3 above)
Comment (DinhNguyenCat) [inline webrtc/libjingle/xmpp/boshsocket.cc:102]: My bad, I forgot to do this. I will update

File: webrtc/libjingle/xmpp/boshsocket.h (hunk 6 above)
Comment (DinhNguyenCat) [inline webrtc/libjingle/xmpp/boshsocket.h]: Noted, I will update using memcpy

**Comment by DinhNguyenCat** [comment]
/cc: @dscherba @willkelleher 
@dscherba I've just updated the code base on your comments.
I also do:
- Add a copy constructor into OutputMsg and InputMsg to avoid some issue like double deallocation during processing with queue.
- Remove typedef (C code style).

Please help to review the PR again.

Thank you,
D.

**Comment by dscherba** [comment]
#8 is tracking this now, closing.

**Pull request closed without merging**